### PR TITLE
`LossVsNoiseAggregator`

### PR DESCRIPTION
**Pull request by AnnaKwa**
It is useful to be able to see plots of loss vs training sample noise to check if the training noise distribution is appropriately set, see how changes in noise-weighted loss affect training, etc. This PR adds that option for downscaling wandb reports but makes it optional since generates a bunch of new 2d figures for each epoch.

- Adds class `LossVsNoiseAggregator`.
- Adds an option to log loss vs noise to wandb via `TrainerConfig.log_loss_vs_noise`. Defaults to False. If true, the new aggregator is used during training.
- adds attributes `ModelOutputs.sigma` (1d tensor of noise used in each training sample) and `ModelOutputs.per_sample_channel_loss` (tensor dict where keys are channel names and values are 1d tensors of loss per sample) so that this information can be passed to the aggregator.

This returns a wandb.Image since wandb's plotly sometimes chokes on plots with two axes and makes the bin counts horizontal bars. 

example output:
[wandb](https://wandb.ai/ai2cm/annak-downscaling/runs/bomshtxf)
image

### CODE REVIEW

File: fme/downscaling/aggregators/main.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import matplotlib.pyplot as plt
 import numpy as np
 import torch
+import wandb
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py]: Instead of importing wandb and sidestepping the FME wrapper interaction, use `WandB.get_instance()` in the aggregator

File: fme/downscaling/aggregators/main.py [hunk 2]
```diff
@@ -50,6 +51,165 @@ def _tensor_mapping_to_numpy(data: TensorMapping) -> TensorMapping:
...
+        self._channel_count[name].scatter_add_(
+            0, bin_indices, torch.ones_like(bin_indices, dtype=torch.int64)
+        )
+
+    @torch.no_grad()
+    def record_batch(self, outputs: ModelOutputs) -> None:
+        if outputs.sigma is None or not outputs.per_sample_channel_loss:
+            return
+
+        sigma = outputs.sigma.detach().flatten().cpu()
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py]: `detach()` should be covered by the no_grad() declaration.

File: fme/downscaling/aggregators/main.py [hunk 3]
```diff
@@ -50,6 +51,165 @@ def _tensor_mapping_to_numpy(data: TensorMapping) -> TensorMapping:
...
+        lines, labels = ax.get_legend_handles_labels()
+        lines2, labels2 = ax2.get_legend_handles_labels()
+        ax.legend(lines + lines2, labels + labels2)
+
+        image = wandb.Image(fig)
+        plt.close(fig)
+        return image
+
+    def get_wandb(self, prefix: str = "") -> Mapping[str, Any]:
+        prefix = ensure_trailing_slash(prefix)
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py]: To simplify the amount of code, I've been implementing (mostly private where possible) `_get` methods which reduce and move the final values to CPU.  That way `get_wandb` directly handles the wandb data prep and handoff and there's less device handling in separate locations.

File: fme/downscaling/aggregators/main.py [hunk 4]
```diff
@@ -50,6 +51,165 @@ def _tensor_mapping_to_numpy(data: TensorMapping) -> TensorMapping:
...
+            color="gray",
+            label="bin count",
+        )
+        ax2.set_ylabel("bin count")
+
+        lines, labels = ax.get_legend_handles_labels()
+        lines2, labels2 = ax2.get_legend_handles_labels()
+        ax.legend(lines + lines2, labels + labels2)
+
+        image = wandb.Image(fig)
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py:168]: e.g., `wandb = WandB.get_instance()` then `wandb.Image(fig)` follows the same standard path.

File: fme/downscaling/aggregators/main.py [hunk 5]
```diff
@@ -50,6 +51,165 @@ def _tensor_mapping_to_numpy(data: TensorMapping) -> TensorMapping:
...
+
+        sigma = outputs.sigma.detach().flatten().cpu()
+        if torch.any(sigma <= 0):
+            raise ValueError("Sigma must be strictly positive for log10 binning")
+        log_sigma = torch.log10(sigma)
+        in_range = (log_sigma >= self._log10_sigma_min) & (
+            log_sigma <= self._log10_sigma_max
+        )
+        if not torch.any(in_range):
+            return
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py:114]: Could move this out into a local helper method `_maybe_get_in_range` that returns `in_range | None` to focus the logic under this method on batch recording.

File: fme/downscaling/aggregators/main.py [hunk 6]
```diff
@@ -50,6 +51,165 @@ def _tensor_mapping_to_numpy(data: TensorMapping) -> TensorMapping:
...
+        lines2, labels2 = ax2.get_legend_handles_labels()
+        ax.legend(lines + lines2, labels + labels2)
+
+        image = wandb.Image(fig)
+        plt.close(fig)
+        return image
+
+    def get_wandb(self, prefix: str = "") -> Mapping[str, Any]:
+        prefix = ensure_trailing_slash(prefix)
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (frodre) [inline fme/downscaling/aggregators/main.py]: Any reason to not use the `fme` function `get_device()`?

File: fme/downscaling/aggregators/main.py (hunk 3 above)
Comment (frodre) [reply fme/downscaling/aggregators/main.py]: I would consider leaving the aggregator values on device until `_get` is called to reduce the amount of moving back and forth.

**Review by frodre — APPROVED** [review]
Thanks.  Overall, LGTM.   The only change I think needs to happen is the usage of `WandB.get_instance` instead of importing `wandb`. I made suggestions to try and improve code readability, but feel free to ignore.
